Getting Started guide for Rails 8

excid3 opened the pull request:
This is the new Rails 8 Getting Started guide. It showcases Rails and provides many introductory points to learn more about all the various features Rails includes like caching, I18n, deploying with Kamal, etc.

This PR also includes an Install Ruby on Rails guide that walks through the basics of installing Ruby with a version manager, arguably one of the biggest barriers to newcomers.

I've also added the ability for highlighting lines in code blocks in the guides. Highlighted lines are specified after the language at the beginning of the code block using a hash and separated by commas. Supports both individual lines and ranges.

```markdown
```ruby#1,3-5,9
```

Live preview available here:
https://excid3-rails-8-getting-start.rails-docs-preview.pages.dev/guides/getting_started
https://excid3-rails-8-getting-start.rails-docs-preview.pages.dev/guides/install_ruby_on_rails

excid3 committed:
Update Getting Started guide for Rails 8
Also introduces Install Ruby on Rails guide

Co-authored-by: Xavier Noria <fxn@hashref.com>
Co-authored-by: Jeremy Daer <jeremydaer@gmail.com>
Co-authored-by: Harriet Oughton <73295547+OughtPuts@users.noreply.github.com>
Co-authored-by: julianduss <julian.duss@renuo.ch>
Co-authored-by: Santiago Rodriguez <46354312+santiagorodriguez96@users.noreply.github.com>
Co-authored-by: Amanda Perino <58528404+AmandaPerino@users.noreply.github.com>
Co-authored-by: Matthew Draper <matthew@trebex.net>
Co-authored-by: Gianlo <gianlorenzo.occhipinti@gmail.com>
Co-authored-by: Collin Jilbert <cjilbert504@gmail.com>

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1548 @@ This guide covers getting up and running with Ruby on Rails.
 * The basic principles of MVC (Model, View, Controller) and RESTful design.
 * How to quickly generate the starting pieces of a Rails application.
-
+* How to deploy your app to production using Kamal.
```
Not sure "How to deploy your app to production" can be part of Getting Started guide.

olimart commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1548 @@ This guide covers getting up and running with Ruby on Rails.
-Creating a New Rails Project
-----------------------------
-
-TIP: You can create new Rails apps with a preconfigured Dev Container development environment. This
-is the fastest way to get started with Rails.
-For instructions see [Getting Started with Dev Containers](getting_started_with_devcontainer.html)
```
given Rails come with `devcontainer` setup, should we not make it the default path to avoid OS-specific issues? 🤷‍♂️ 
(or at least keep the mention)

excid3 replied:
I think we'll move this to the Install Ruby on Rails guide.

AmandaPerino replied:
Makes sense in the Install guide.

excid3 replied:
Added in 8f4918d4f0

excid3 committed:
Fix linter issues

andyw8 commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,131 @@
+Installing Rails
+------------------------
+
+A "gem" in Ruby is a self-contained package of a library or Ruby program. We can use Ruby's `gem` command to install the latest version of Rails and its dependencies from [RubyGems.org](https://rubygems.org).
```
```suggestion
A "gem" in Ruby is a self-contained package of a library or Ruby program. We can use Bundler's `gem` command to install the latest version of Rails and its dependencies from [RubyGems.org](https://rubygems.org).
```
(although the line is blurred since Bundler is built-in)

excid3 replied:
We'd probably need to explain Bundler here too?

andyw8 replied:
Yeah, it might be too much for an intro.

p8 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+It works!
 
-Let's take a look at the contents of our new migration file:
+Active Record Model Basics
+-------------------------
```
```suggestion
Active Record Model Basics
--------------------------
```

p8 commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
 
-```html+erb
-<h1>Articles</h1>
+A Request's Journey Through Rails
+---------------------
````
```suggestion
A Request's Journey Through Rails
---------------------------------
```

p8 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+
+In `product_params`, we tell Rails to inspect the params and ensure there is a key named `:product` with an array of parameters as the value. The only permitted parameters for products is `:name` and Rails will ignore any other parameters. This protects our application from malicious users who might try to hack our application.
+
+#### Handling errors
```
```suggestion
#### Handling Errors
```

rjrobinson commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
 
+### Model-View-Controller Basics
+
+Rails code is organized using the Model-View-Controller (MVC) architecture. With MVC, we have three main concepts where the majority of our code lives:
```
We could deep link to that pattern: 

https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller 

or 

https://www.w3schools.in/mvc-architecture

AmandaPerino replied:
In general we try not to link out as much as possible (not always feasible), but we are creating a diagram for this section to help visualize MVC architecture, which should help folks understand it better (and keep them on the page). It should go live next week with the tutorial.

rjrobinson replied:
Nice! That makes sense.

p8 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+
+Learn more in the [Asset Pipeline](asset_pipeline.html) and [Working with JavaScript in Rails](working_with_javascript_in_rails.html) guides.
+
+Rich Text Fields with Action Text
+--------------------------------
```
```suggestion
Rich Text Fields with Action Text
---------------------------------
```

p8 commented on guides/source/getting_started.md:
```diff
@@ -1159,987 +1560,827 @@ the following contents:
 
-TIP: For more information on routing, see the [Rails Routing](routing.html)
-guide.
+### In stock email notifications
```
```suggestion
### In Stock Email Notifications
```

p8 commented on guides/source/getting_started.md:
```diff
@@ -1159,987 +1560,827 @@ the following contents:
+`notify_subscribers` uses the Active Record association to query the `subscribers` table for all subscribers for this specific product and then queues up the `in_stock` email to be sent to each of them.
 
-You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog`, two folders were created within `app/` along with the rest:
+### Extracting A Concern
```
```suggestion
### Extracting a Concern
```

p8 commented on guides/source/getting_started.md:
```diff
@@ -1159,987 +1560,827 @@ the following contents:
-</div>
+Learn more about [Testing Rails Applications](testing.html)
+
+Consistently formatted code with Rubocop
```
```suggestion
Consistently Formatted Code with Rubocop
```

p8 commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,131 @@
+
+--------------------------------------------------------------------------------
+
+## Choose your operating system
```
```suggestion
## Choose Your Operating System
```

p8 commented on guides/source/install_ruby_on_rails.md:
````diff
@@ -0,0 +1,131 @@
+$ mise use -g ruby@3
+```
+
+Verifying your Ruby install
````
```suggestion
Verifying Your Ruby Install
```

p8 commented on guides/source/install_ruby_on_rails.md:
````diff
@@ -0,0 +1,131 @@
+ruby 3.3.6
+```
+
+Installing Rails
+------------------------
````
```suggestion
Installing Rails
----------------
```

p8 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+
+Log out and visit the products index and show pages to see they're accessible without being authenticated.
+
+### Showing Links For Authenticated Users Only
```
```suggestion
### Showing Links for Authenticated Users Only
```

p8 reviewed (commented):
Very nice rewrite!
I added some nitpicks about headings.
https://guides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#headings

andyw8 commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,131 @@
+
+### Install Ruby on Windows
+
+The Windows Subsystem for Linux will provide the best experience for Ruby on Rails development on Windows. It runs Ubuntu inside of Windows which allows you to work in an environment that is close to what your servers will run in production.
```
```suggestion
The Windows Subsystem for Linux (WSL) will provide the best experience for Ruby on Rails development on Windows. It runs Ubuntu inside of Windows which allows you to work in an environment that is close to what your servers will run in production.
```

azzenabidi commented:
> rails --version
> Rails 8.1.0

Are you planning to release this guide by the time 8.1.0 is out? either way I suggest you add: "at the time of writing this guide, rails is version x "

In the introduction, it would be great to link to the install rails guide as well. 

I think it would be more beneficial if the CI/CD with github actions comes after the section deploy to production so that there is more explanation how to integrate kamal with github actions for faster deployment. for context, when I deployment with kamal I was a bit disappointment coz it took 30 minutes - not the 80 seconds of the demo - deploaying with github action is certainly much faster.

p8 commented:
> > rails --version
> > Rails 8.1.0
> 
> Are you planning to release this guide by the time 8.1.0 is out? either way I suggest you add: "at the time of writing this guide, rails is version x "

This will be backported to 8-0-stable, so it will be released when the next version of 8.0 is released.

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1548 @@ This guide covers getting up and running with Ruby on Rails.
+
+### Running Migrations
+
+Now that you have defined what change to make to the database, use the following command to run the migrations:
```
Now that you have defined what changes to make to the database

excid3 committed:
Apply suggestions from code review

Co-authored-by: Petrik de Heus <petrik@deheus.net>
Co-authored-by: Andy Waite <github.aw@andywaite.com>

Youxaf-khan commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1548 @@ This guide covers getting up and running with Ruby on Rails.
-  <% end %>
-</ul>
-```
+When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will the request and return a response to the browser.
````
When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will map the request and returns a response to the browser.

excid3 committed:
Add Dev Container to Install guide
Also updates Dev Container guide to reference the Store application.

Co-authored-by: Olivier <olimart@users.noreply.github.com>

excid3 committed:
Use range for highlighting lines

olimart commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
 
-Let's start by adding a route to our routes file, `config/routes.rb`, at the
-top of the `Rails.application.routes.draw` block:
+You may also notice that Rails applications rarely use `require` statements. Rails uses naming conventions to require files automatically so you can focus on writing your application code.
```
what if your background is completely different, maybe you don't know how Ruby works either, therefore, spotting the difference may not be obvious.
I would keep this out, or move to the app folder structure and mention that Rails strongly recommends sticking to naming conventions to avoid `require`.

AmandaPerino replied:
Or: "You may also notice that Rails applications rarely use `require` statements like you may have seen in other programming languages." 
(I swear we had something like this before.)

excid3 replied:
Added @AmandaPerino's line. I think we had that earlier too but it must have gotten lost somewhere.

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1548 @@ This guide covers getting up and running with Ruby on Rails.
+Check out the [Action Text Overview](action_text_overview.html) to learn more.
 
-#### Using Partials to Share View Code
+File Uploads with Active Storage
+-------------------------------
```
File Uploads with Active Storage
--------------------------------

Youxaf-khan replied:
File Uploads with Active Storage
--------------------------------

Youxaf-khan replied:
File Uploads with Active Storage
#### --------------------------------

Youxaf-khan replied:
missing a hyphen here

Youxaf-khan commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,133 @@
+**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**
+
+Install Ruby on Rails Guide
+=====================
```
```suggestion
Install Ruby on Rails Guide
===========================
```

Youxaf-khan commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,131 @@
+
+A "gem" in Ruby is a self-contained package of a library or Ruby program. We can use Ruby's `gem` command to install the latest version of Rails and its dependencies from [RubyGems.org](https://rubygems.org).
+
+Run the following to install the latest Rails and make it available in your terminal:
```
@excid3 Thanks a lot, Will use it form now on

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+### Running Migrations
 
-You only need `require` calls for two use cases:
+Now that you have defined what change to make to the database, use the following command to run the migrations:
```
```suggestion
Now that you have defined what changes to make to the database, use the following command to run the migrations:
```

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
-https://en.wikipedia.org/wiki/Create,_read,_update,_and_delete) operations. You
-may even find that the majority of the work your application does is CRUD. Rails
-acknowledges this, and provides many features to help simplify code doing CRUD.
+This deleted the T-Shirt product from our database. We can confirm this with `Product.all` to see that that it only returns Pants.
```
```suggestion
This deleted the T-Shirt product from our database. We can confirm this with `Product.all` to see that it only returns Pants.
```

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
-  def new
-    @article = Article.new
-  end
+When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will the request and return a response to the browser.
```
```suggestion
When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will map/process the request and return a response to the browser.
```

Youxaf-khan commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+Check out the [Action Text Overview](action_text_overview.html) to learn more.
 
-#### Using Partials to Share View Code
+File Uploads with Active Storage
+-------------------------------
```
```suggestion
File Uploads with Active Storage
--------------------------------
```

Youxaf-khan commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,133 @@
+
+A "gem" in Ruby is a self-contained package of a library or Ruby program. We can use Ruby's `gem` command to install the latest version of Rails and its dependencies from [RubyGems.org](https://rubygems.org).
+
+Run the following to install the latest Rails and make it available in your terminal:
```
```suggestion
Run the following command to install the latest Rails and make it available in your terminal:
```

elhalvers commented:
Hey Chris! Still working through the new guide. I really like using the Console under '7. Active Record Model Basics' to do the CRUD actions and validations before introducing the Controller/Views!

One thing, I noticed so far while working through this section - 
After adding the validation on the Product model, I had to use 'reload!' in my console before testing the validation. I am running the console on a different tab in the same window so before reloading the console, it let me save the new record without the name so if you haven't already, might want to mention that 😊

avinashjoshi commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+To stop the Rails server anytime, press `Ctrl-C` in your terminal.
 
-### Say "Hello", Rails
+### Automatic Reloading in Development
```
I got a bit excited and thought the exact same thing!

excid3 replied:
I've renamed this to `Autoloading in Development` to better clarify that.

jespr commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
+<div id="products">
+  <% @products.each do |product| %>
+    <div>
+        <%= link_to product.name, product %>
```
Looks like 4 space indentation here instead of just 2 :)

excid3 replied:
Good catch! We had a mix of tabs and spaces from earlier reviews, so this probably left over from converting those.

zzak commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,133 @@
+
+This guide will walk you through installing the Ruby programming language and the Rails framework on your operating system.
+
+While your OS might come with Ruby pre-installed, it's often outdated and can't be upgraded. Using a version manager like [Mise](https://mise.jdx.dev/getting-started.html) allows you to install the latest Ruby version, use a different Ruby version for each app, and easily upgrade to new versions when they're released.
```
I've never heard of mise until today, why not use something most Ruby developers use or are familiar with like `rbenv`?

excid3 replied:
This is what's used in Omakub and what DHH approved to use here.

khhizr007 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1551 @@ This guide covers getting up and running with Ruby on Rails.
-tremendous increase in productivity. If you persist in bringing old habits from
-other languages to your Rails development, and trying to use patterns you
-learned elsewhere, you may have a less happy experience.
+Rails is opinionated software. It makes the assumption that there is a "best" way to do things, and it's designed to encourage that way - and in some cases to discourage alternatives. If you learn "The Rails Way" you'll probably discover a tremendous increase in productivity. If you persist in bringing old habits from other languages to your Rails development, and trying to use patterns you learn ...
```
```suggestion
Rails is an opinionated software. It makes the assumption that there is a "best" way to do things, and it's designed to encourage that way - and in some cases to discourage alternatives. If you learn "The Rails Way" you'll probably discover a tremendous increase in your productivity. If you persist in bringing old habits from other languages to your Rails development, and try to use patterns you learned elsewhere, you may have a less happy experience.
```

excid3 replied:
Since this is what's in the current guide, I'm going to leave it as-is.

p8 commented:
> I've also added the ability for highlighting lines in code blocks in the guides. Highlighted lines are specified after the language at the beginning of the code block using a hash and separated by commas. Supports both individual lines and ranges.
> 
> ```
> ```ruby#1,3-5,9
> ```
> 
> Live preview available here: https://0e60ea9a.rails-docs-preview.pages.dev/guides/getting_started https://0e60ea9a.rails-docs-preview.pages.dev/guides/install_ruby_on_rails

Maybe I'm missing something (or will this require additional CSS?), but I'm not seeing highlighted lines:
<img width="862" alt="image" src="https://github.com/user-attachments/assets/42a1c178-10a1-4729-9f27-78424f9b0894">

AmandaPerino commented:
> Maybe I'm missing something (or will this require additional CSS?), but I'm not seeing highlighted lines: 

Indeed, we are still working on the highlight styling. Chris added the ability to highlight by calling out the lines, but the highlight itself is still coming.

excid3 committed:
Apply suggestions from code review

Co-authored-by: M. Yousaf Khan <79319679+Youxaf-khan@users.noreply.github.com>

cappe commented:
Great guide! I found something that I felt a bit unintuitive. In https://0e60ea9a.rails-docs-preview.pages.dev/guides/getting_started#unsubscribe-links, `#show` is being used to destroy a subscription. Would it make more sense to use, well, `#destroy`?

excid3 commented:
@cappe you can't make a DELETE request from a link in an email. Has to be a GET.

excid3 committed:
Run reload! in console after adding validation

Co-authored-by: Eric Halverson <74928397+elhalvers@users.noreply.github.com>

excid3 committed:
Clarify a few sections

cappe commented:
@excid3 Oh that figures. Excellent work!

excid3 commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,133 @@
+
+While your OS might come with Ruby pre-installed, it's often outdated and can't be upgraded. Using a version manager like [Mise](https://mise.jdx.dev/getting-started.html) allows you to install the latest Ruby version, use a different Ruby version for each app, and easily upgrade to new versions when they're released.
+
+Alternatively, you can use Dev Containers to run Rails without installing Ruby or Rails directly on your machine. Check out the [Getting Started with Dev Containers](getting_started_with_devcontainer.html) guide to learn more.
```
The current guides do not have an Install Ruby on Rails guide, so it got extra emphasis. Now that we're introducing an Install guide, we want it to be part of the intro and readers can choose what works best for them.

MatheusRich commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -160,6 +176,21 @@ def api_link(url)
           end
         end
+
+        # Parses "ruby#3,5-6,10" into ["ruby", [3,5,6,10]] for highlighting line numbers in code blocks
```
I'm curious about the syntax here. Is it something you came up with, or got from somewhere?

The CommonMark spec [does mention "info strings"](https://spec.commonmark.org/0.30/#example-143) (I'm not 100% sure about their purpose). They look like metadata, which could be handy here.

excid3 replied:
Thanks for sharing, hadn't seen this. I think using `ruby highlight=3,5-6,10` might be a better format like examples in the spec. Since they suggest the first word as the language, separating with a space instead of hash would fit better. Will refactor.

excid3 replied:
Actually, seems like Redcarpet only passes the first word. I remember hitting that previously. Redcarpet's a C extension so I don't think that's gonna be an easy one.

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+  end
+
+  private
+
```
```suggestion
```

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+  end
+
+  private
+
```
```suggestion
```

rafaelfranca replied:
The generator, and our styleguide doesn't have an empty line after the `private`

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+  end
+
+  private
+
```
```suggestion
```

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+  end
+
+  private
+
```
```suggestion
```

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+
+By passing `@product` into `cache`, Rails generates a unique cache key for the product. Active Record objects have a `cache_key` method that returns a String like `"products/1"`. The `cache` helper in the views combines this with the template digest to create a unique key for this HTML.
+
+To enable caching in development, run `rails dev:cache` in your terminal.
```
```suggestion
To enable caching in development, run `bin/rails dev:cache` in your terminal.
```

I didn't check everything but we should keep the commands consistent

rafaelfranca commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+
+```ruby#2
+class Product
+  has_rich_text :description
````
Where did we create the description column? Would this work without the column?

excid3 replied:
This is using ActionText, so it doesn't need it's own column since it uses the ActionText tables.

excid3 committed:
Apply suggestions from code review

Co-authored-by: Rafael Mendonça França <rafael@franca.dev>

excid3 committed:
Use bin/rails consistently

austernotus commented:
Lovely work everyone!
Have you considered paginating the guide? 

Feels a bit too overwhelming at the moment (I know there are chapters at the side but it's not quite the same)

excid3 commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1558 @@ This guide covers getting up and running with Ruby on Rails.
+    if @product.update(product_params)
+      redirect_to @product
+    else
+      render :new, status: :unprocessable_entity
```
Good catch, should be `:edit`. 

Want to use the Suggestion feature to change it?

siasmj replied:
```suggestion
      render :edit, status: :unprocessable_entity
```

siasmj commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+    if @product.update(product_params)
+      redirect_to @product
+    else
+      render :new, status: :unprocessable_entity
```
```suggestion
      render :edit, status: :unprocessable_entity
```

siasmj commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+    if @product.update(product_params)
+      redirect_to @product
+    else
+      render :new, status: :unprocessable_entity
```
```suggestion
      render :edit, status: :unprocessable_entity
```

siasmj reviewed (commented):
I found a couple more instances where `:edit` should be rendered rather than `:new`.

excid3 committed:
Apply suggestions from code review

Co-authored-by: Matt Sias <mattsias@gmail.com>

aesmail commented:
@excid3 amazing! Any reason why `bin/rails server` is used over `bin/dev`?

Schwad commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+We're going to build a project called `store` - a simple e-commerce app that demonstrates several of Rails' built-in features.
 
-#### Installing SQLite3
+TIP: Any commands prefaced with a dollar sign `$` should be run in the terminal.
```
<img width="677" alt="image" src="https://github.com/user-attachments/assets/f8f07eab-9306-4b28-940e-be7a49b2fe02">

Taking this copy, making it a TIP and placing it right after "Creating a New Rails App" is such a GREAT UX.

It's very easy for us to remember this but it makes the guide accessible to people at the "Hello World" level. Love it.

AmandaPerino commented on guides/source/getting_started_with_devcontainer.md:
```diff
@@ -60,8 +60,8 @@ take care of installing the correct Ruby and Rails versions for you.
 
-Creating the Blog Application
------------------------------
+Creating the Store Application
```
Hi @duleorlovic - this is a tutorial for a completely different app. This one walks you through creating an e-commerce app, hence the `store` name.

vipulnsward replied:
It would be great to replace these references out to use the store app models/controllers in future, once this PR merged, for consistency across guides. 

Ex this guide refers to blog example, which can feel out of context or in some places we refer to existing knowledge or "Blog related models, setup , etc"

https://github.com/rails/rails/blob/5608e7b5e4e71e39c8dee31253820fe0955ec494/guides/source/debugging_rails_applications.md 

Not in the scope of this PR though.

excid3 committed:
Support <picture> element in guides

vipulnsward commented on guides/source/install_ruby_on_rails.md:
```diff
@@ -0,0 +1,133 @@
+$ source ~/.zshrc
+$ brew install openssl@3 libyaml gmp rust
+
+# Install Mise version manager
```
Question about this choice: (And also this guide): 
Ruby comes with this guide itself: https://www.ruby-lang.org/en/documentation/installation/#managers , I am assuming we wanted to give user a simpler/easier way to follow to install Ruby?

Why not prefer mentioning most popular Ruby Version managers from the community: https://railsdeveloper.com/survey/2024/#ruby-rails-version-updates which have better support/community around it for Ruby use?

vipulnsward commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
 ```
 
-If it says something like "Rails 8.1.0", you are ready to continue.
+The version shown should be Rails 8.0.0 or higher.
````
```suggestion
The version shown should be Rails 8.1.0 or higher.
```

rafaelfranca replied:
Keep 8.0.0 in all of those. We can change to 8.1.0 after it is released

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
-
-#### Installing Rails
+* Ruby 3.2 or newer
+* Rails 8.0.0 or newer
```
```suggestion
* Rails 8.1.0 or newer
```

vipulnsward commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+
+```bash
+=> Booting Puma
+=> Rails 8.0.0 application starting in development
````
```suggestion
=> Rails 8.1.0 application starting in development
```

rafaelfranca commented on guides/source/getting_started.md:
```diff
@@ -1159,987 +1563,827 @@ the following contents:
-the [Ruby on Rails Security Guide](security.html).
+Solid Queue is enabled for our production Kamal deployment using the `SOLID_QUEUE_IN_PUMA: true` environment variable to `config/deploy.yml`. This tells our web server, Puma, to start and stop the Solid Queue process automatically.
 
+When emails are sent with Action Mailer's `deliver_later`, these emails will be sent to ActiveJob for sending in the background so they don't delay the HTTP request. With Solid Queue in production, emails will be sent in the background, automatically retried if they fail to send, and jobs are kept safe in the database during restarts.
```
```suggestion
When emails are sent with Action Mailer's `deliver_later`, these emails will be sent to Active Job for sending in the background so they don't delay the HTTP request. With Solid Queue in production, emails will be sent in the background, automatically retried if they fail to send, and jobs are kept safe in the database during restarts.
```

vipulnsward commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
-### Database Migrations
+```irb
+store(dev)> Rails.version
+=> "8.0.0"
````
```suggestion
=> "8.1.0"
```

vipulnsward commented on guides/assets/images/getting_started/rails_welcome.png:
Sorry about the extra work here. Can we regenerate this one for 8.1.0(current dev version)?

rafaelfranca replied:
This will be released in next 8.0. Let's keep this pr as is, and change after it is released

vipulnsward replied:
Cool, in that case need to revert some 8.1.0 uses back to 8.0.0 in this guide:

![image](https://github.com/user-attachments/assets/2520076a-8980-49ba-8c75-bf8f25f14342)

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+This command does several things. It creates...
 
-Let's open `app/views/articles/index.html.erb`, and replace its contents with:
+1. a migration in the `db/migrate` folder
```
```suggestion
1. A migration in the `db/migrate` folder.
```

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
 
-Let's open `app/views/articles/index.html.erb`, and replace its contents with:
+1. a migration in the `db/migrate` folder
+2. an Active Record model in `app/models/product.rb`
```
```suggestion
2. An Active Record model in `app/models/product.rb`.
```

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
-Let's open `app/views/articles/index.html.erb`, and replace its contents with:
+1. a migration in the `db/migrate` folder
+2. an Active Record model in `app/models/product.rb`
+3. tests and test fixtures for this model
```
```suggestion
3. Tests and test fixtures for this model.
```

vipulnsward commented on guides/source/getting_started.md:
```diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
+Now that we have created our products table, we can interact with it in Rails. Let's try it out.
 
-To define a model, we will use the model generator:
+For this, we're going to use a feature of Rails called the *console*.  The console is a helpful, interactive tool for testing our code in our Rails application.
```
```suggestion
For this, we're going to use a Rails feature called the *console*. The console is a helpful, interactive tool for testing our code in our Rails application.
```

vipulnsward commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
-create      test/models/article_test.rb
-create      test/fixtures/articles.yml
+```irb
+Loading development environment (Rails 8.1.0)
````
nit: version

excid3 replied:
```suggestion
Loading development environment (Rails 8.0.0)
```

vipulnsward commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
 ```
 
-At this prompt, we can initialize a new `Article` object:
+Similar to `save`, we can use `create` to instantiate and save an Active Record model in a single call.
````
```suggestion
Similar to `save`, we can use `create` to instantiate and save an Active Record object in a single call.
```

excid3 commented on guides/source/getting_started.md:
````diff
@@ -7,1150 +7,1554 @@ This guide covers getting up and running with Ruby on Rails.
 
 ```bash
 $ rails --version
 Rails 8.1.0
````
```suggestion
Rails 8.0.0
```

excid3 committed:
Apply suggestions from code review

Co-authored-by: Vipul A M <vipul@hey.com>
Co-authored-by: Rafael Mendonça França <rafael@franca.dev>

p8 commented:
Can you wrap this guide at 80 columns?
See: https://edgeguides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#column-wrapping

excid3 committed:
Revert "Support <picture> element in guides"

This reverts commit 9fe84c737b5649798787e6787981b57948820341.

jathayde committed:
Adding diagrams in getting started

jathayde committed:
code highlighting for .hll (new lines added)

jathayde committed:
a bit of spacing on the highlight border callout

excid3 committed:
Add more code highlights and make a few things more consistent

excid3 committed:
Merge branch 'rails-8-getting-started-guide' of github.com:jathayde/rails into rails-8-getting-started-guide

excid3 committed:
80 character line limit

excid3 committed:
Add styles for product show page

This also moves CSS & JavaScript section until after the features are built so styling can be applied all at once.

Co-authored-by: John Athayde <jmpa@meticulous.com>

elhalvers commented on guides/source/getting_started.md:
```diff
@@ -1151,11 +1400,15 @@ end
 
-We also want to replace any instance variables with a local variable, which we can define when we render the partial. We'll replace `@product` with `product`.
+We also want to replace any instance variables with a local variable, which we
+can define when we render the partial. We'll replace `@product` with `product`.
```
Might be nice to give a brief explanation, for example:

We also want to replace any instance variables with a local variable so the partial can be used with different views and display different data that is passed to it.

This can be defined when we render the partial. We'll replace `@product` with `product`.

excid3 replied:
Thanks @elhalvers. We discussed this a while back and didn't have a great explanation at the time. I can run with your suggestion. 👍

excid3 committed:
Explain local variables for partials

Co-authored-by: Eric Halverson <74928397+elhalvers@users.noreply.github.com>

excid3 committed:
Update text

excid3 committed:
Adjust to 80 chars

maxandygit commented:
I really appreciate the new tutorial!

Not sure if someone suggest that already! but I was wondering if it might be more helpful to display the error messages directly in the browser (in addition to the console) after adding the validation in the product model. This could make it easier for beginners to understand what’s happening.

For reference, here’s an example from the official documentation: https://guides.rubyonrails.org/getting_started.html#validations-and-displaying-error-messages
